File: Samples/TableEditor/CS/Form1.cs
// Copyright (C) 2004-2005 MySQL AB
//
// MySQL Connector/NET is licensed under the terms of the GPLv2
// <http://www.gnu.org/licenses/old-licenses/gpl-2.0.html>, like most 
// MySQL Connectors. There are special exceptions to the terms and 
// conditions of the GPLv2 as it is applied to this software, see the 
// FLOSS License Exception
// <http://www.mysql.com/about/legal/licensing/foss-exception.html>.
//
// This program is free software; you can redistribute it and/or modify 
// it under the terms of the GNU General Public License as published 
// by the Free Software Foundation; version 2 of the License.
//
// This program is distributed in the hope that it will be useful, but 
// WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY 
// or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License 
// for more details.
//
// You should have received a copy of the GNU General Public License along 
// with this program; if not, write to the Free Software Foundation, Inc., 
// 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA

using System;
using System.Drawing;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using System.Data;
using MySql.Data.MySqlClient;

namespace TableEditor
{
	/// <summary>
	/// Summary description for Form1.
	/// </summary>
	public class Form1 : System.Windows.Forms.Form
	{
		private System.Windows.Forms.Label label1;
		private System.Windows.Forms.Label label2;
		private System.Windows.Forms.Label label3;
		private System.Windows.Forms.Label label4;
		private System.Windows.Forms.ComboBox tables;
		private System.Windows.Forms.TextBox server;
		private System.Windows.Forms.TextBox userid;
		private System.Windows.Forms.TextBox password;
		private System.Windows.Forms.Button connectBtn;
		private System.Windows.Forms.Button updateBtn;
		/// <summary>
		/// Required designer variable.
		/// </summary>
		private System.ComponentModel.Container components = null;
		private System.Windows.Forms.ComboBox databaseList;
		private System.Windows.Forms.Label label5;

		private MySqlConnection		conn;
		private DataTable			data;
		private MySqlDataAdapter	da;
		private System.Windows.Forms.DataGrid dataGrid;
		private MySqlCommandBuilder	cb;

		public Form1()
		{
			//
			// Required for Windows Form Designer support
			//
			InitializeComponent();

			//
			// TODO: Add any constructor code after InitializeComponent call
			//
		}

		/// <summary>
		/// Clean up any resources being used.
		/// </summary>
		protected override void Dispose( bool disposing )
		{
			if( disposing )
			{
				if (components != null) 
				{
					components.Dispose();
				}
			}
			base.Dispose( disposing );
		}

		#region Windows Form Designer generated code
		/// <summary>
		/// Required method for Designer support - do not modify
		/// the contents of this method with the code editor.
		/// </summary>
		private void InitializeComponent()
		{
			this.label1 = new System.Windows.Forms.Label();
			this.server = new System.Windows.Forms.TextBox();
			this.userid = new System.Windows.Forms.TextBox();
			this.label2 = new System.Windows.Forms.Label();
			this.password = new System.Windows.Forms.TextBox();
			this.label3 = new System.Windows.Forms.Label();
			this.connectBtn = new System.Windows.Forms.Button();
			this.label4 = new System.Windows.Forms.Label();
			this.tables = new System.Windows.Forms.ComboBox();
			this.dataGrid = new System.Windows.Forms.DataGrid();
			this.updateBtn = new System.Windows.Forms.Button();
			this.databaseList = new System.Windows.Forms.ComboBox();
			this.label5 = new System.Windows.Forms.Label();
			((System.ComponentModel.ISupportInitialize)(this.dataGrid)).BeginInit();
			this.SuspendLayout();
			// 
			// label1
			// 
			this.label1.Location = new System.Drawing.Point(8, 11);
			this.label1.Name = "label1";
			this.label1.Size = new System.Drawing.Size(48, 16);
			this.label1.TabIndex = 0;
			this.label1.Text = "Server:";
			// 
			// server
			// 
			this.server.Location = new System.Drawing.Point(56, 8);
			this.server.Name = "server";
			this.server.Size = new System.Drawing.Size(320, 20);
			this.server.TabIndex = 1;
			this.server.Text = "";
			// 
			// userid
			// 
			this.userid.Location = new System.Drawing.Point(56, 32);
			this.userid.Name = "userid";
			this.userid.Size = new System.Drawing.Size(120, 20);
			this.userid.TabIndex = 3;
			this.userid.Text = "";
			// 
			// label2
			// 
			this.label2.Location = new System.Drawing.Point(8, 37);
			this.label2.Name = "label2";
			this.label2.Size = new System.Drawing.Size(48, 16);
			this.label2.TabIndex = 2;
			this.label2.Text = "User Id:";
			// 
			// password
			// 
			this.password.Location = new System.Drawing.Point(260, 32);
			this.password.Name = "password";
			this.password.PasswordChar = '*';
			this.password.Size = new System.Drawing.Size(116, 20);
			this.password.TabIndex = 5;
			this.password.Text = "";
			// 
			// label3
			// 
			this.label3.Location = new System.Drawing.Point(192, 37);
			this.label3.Name = "label3";
			this.label3.Size = new System.Drawing.Size(56, 16);
			this.label3.TabIndex = 4;
			this.label3.Text = "Password:";
			// 
			// connectBtn
			// 
			this.connectBtn.Location = new System.Drawing.Point(400, 8);
			this.connectBtn.Name = "connectBtn";
			this.connectBtn.TabIndex = 6;
			this.connectBtn.Text = "Connect";
			this.connectBtn.Click += new System.EventHandler(this.connectBtn_Click);
			// 
			// label4
			// 
			this.label4.Location = new System.Drawing.Point(8, 109);
			this.label4.Name = "label4";
			this.label4.Size = new System.Drawing.Size(64, 16);
			this.label4.TabIndex = 0;
			this.label4.Text = "Tables";
			// 
			// tables
			// 
			this.tables.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
			this.tables.Location = new System.Drawing.Point(80, 104);
			this.tables.Name = "tables";
			this.tables.Size = new System.Drawing.Size(296, 21);
			this.tables.TabIndex = 7;
			this.tables.SelectedIndexChanged += new System.EventHandler(this.tables_SelectedIndexChanged);
			// 
			// dataGrid
			// 
			this.dataGrid.Anchor = ((System.Windows.Forms.AnchorStyles)((((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Bottom) 
				| System.Windows.Forms.AnchorStyles.Left) 
				| System.Windows.Forms.AnchorStyles.Right)));
			this.dataGrid.DataMember = "";
			this.dataGrid.HeaderForeColor = System.Drawing.SystemColors.ControlText;
			this.dataGrid.Location = new System.Drawing.Point(8, 136);
			this.dataGrid.Name = "dataGrid";
			this.dataGrid.Size = new System.Drawing.Size(544, 312);
			this.dataGrid.TabIndex = 8;
			// 
			// updateBtn
			// 
			this.updateBtn.Anchor = ((System.Windows.Forms.AnchorStyles)((System.Windows.Forms.AnchorStyles.Top | System.Windows.Forms.AnchorStyles.Right)));
			this.updateBtn.Location = new System.Drawing.Point(477, 104);
			this.updateBtn.Name = "updateBtn";
			this.updateBtn.TabIndex = 9;
			this.updateBtn.Text = "Update";
			this.updateBtn.Click += new System.EventHandler(this.updateBtn_Click);
			// 
			// databaseList
			// 
			this.databaseList.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList;
			this.databaseList.Location = new System.Drawing.Point(80, 80);
			this.databaseList.Name = "databaseList";
			this.databaseList.Size = new System.Drawing.Size(296, 21);
			this.databaseList.TabIndex = 11;
			this.databaseList.SelectedIndexChanged += new System.EventHandler(this.databaseList_SelectedIndexChanged);
			// 
			// label5
			// 
			this.label5.Location = new System.Drawing.Point(8, 85);
			this.label5.Name = "label5";
			this.label5.Size = new System.Drawing.Size(64, 16);
			this.label5.TabIndex = 10;
			this.label5.Text = "Databases";
			// 
			// Form1
			// 
			this.AutoScaleBaseSize = new System.Drawing.Size(5, 13);
			this.ClientSize = new System.Drawing.Size(560, 453);
			this.Controls.Add(this.databaseList);
			this.Controls.Add(this.label5);
			this.Controls.Add(this.updateBtn);
			this.Controls.Add(this.dataGrid);
			this.Controls.Add(this.tables);
			this.Controls.Add(this.connectBtn);
			this.Controls.Add(this.password);
			this.Controls.Add(this.label3);
			this.Controls.Add(this.userid);
			this.Controls.Add(this.label2);
			this.Controls.Add(this.server);
			this.Controls.Add(this.label1);
			this.Controls.Add(this.label4);
			this.Name = "Form1";
			this.Text = "Form1";
			((System.ComponentModel.ISupportInitialize)(this.dataGrid)).EndInit();
			this.ResumeLayout(false);

		}
		#endregion

		/// <summary>
		/// The main entry point for the application.
		/// </summary>
		[STAThread]
		static void Main() 
		{
			Application.Run(new Form1());
		}

		private void connectBtn_Click(object sender, System.EventArgs e)
		{
			if (conn != null)
				conn.Close();
	
			string connStr = String.Format("server={0};user id={1}; password={2}; database=mysql; pooling=false",
				server.Text, userid.Text, password.Text );

			try 
			{
				conn = new MySqlConnection( connStr );
				conn.Open();

				GetDatabases();
			}
			catch (MySqlException ex) 
			{
				MessageBox.Show( "Error connecting to the server: " + ex.Message );
			}
		}

		private void GetDatabases() 
		{
			MySqlDataReader reader = null;

			MySqlCommand cmd = new MySqlCommand("SHOW DATABASES", conn);
			try 
			{
				reader = cmd.ExecuteReader();
				databaseList.Items.Clear();
				while (reader.Read()) 
				{
					databaseList.Items.Add( reader.GetString(0) );
				}
			}
			catch (MySqlException ex) 
			{
				MessageBox.Show("Failed to populate database list: " + ex.Message );
			}
			finally 
			{
				if (reader != null) reader.Close();
			}
		}

		private void databaseList_SelectedIndexChanged(object sender, System.EventArgs e)
		{
			MySqlDataReader reader = null;

			conn.ChangeDatabase( databaseList.SelectedItem.ToString() );

			MySqlCommand cmd = new MySqlCommand("SHOW TABLES", conn);
			try 
			{
				reader = cmd.ExecuteReader();
				tables.Items.Clear();
				while (reader.Read()) 
				{
					tables.Items.Add( reader.GetString(0) );
				}
			}
			catch (MySqlException ex) 
			{
				MessageBox.Show("Failed to populate table list: " + ex.Message );
			}
			finally 
			{
				if (reader != null) reader.Close();
			}
		}

		private void tables_SelectedIndexChanged(object sender, System.EventArgs e)
		{
			data = new DataTable();
			
			da = new MySqlDataAdapter("SELECT * FROM " + tables.SelectedItem.ToString(), conn );
			cb = new MySqlCommandBuilder( da );

			da.Fill( data );

			dataGrid.DataSource = data;
		}

		private void updateBtn_Click(object sender, System.EventArgs e)
		{
			DataTable changes = data.GetChanges();
			da.Update( changes );
			data.AcceptChanges();
		}
	}
}

File: Samples/TableEditor/VB/Form1.vb
' Copyright (C) 2004-2005 MySQL AB
'
' This program is free software; you can redistribute it and/or modify
' it under the terms of the GNU General Public License version 2 as published by
' the Free Software Foundation
'
' There are special exceptions to the terms and conditions of the GPL 
' as it is applied to this software. View the full text of the 
' exception in file EXCEPTIONS in the directory of this software 
' distribution.
'
' This program is distributed in the hope that it will be useful,
' but WITHOUT ANY WARRANTY; without even the implied warranty of
' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
' GNU General Public License for more details.
'
' You should have received a copy of the GNU General Public License
' along with this program; if not, write to the Free Software
' Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA 

Imports System
Imports System.Data
Imports System.Windows.Forms
Imports MySql.Data.MySqlClient

Public Class Form1
    Inherits System.Windows.Forms.Form

    Dim conn As MySqlConnection
    Dim data As DataTable
    Dim da As MySqlDataAdapter
    Dim cb As MySqlCommandBuilder


#Region " Windows Form Designer generated code "

    Public Sub New()
        MyBase.New()

        'This call is required by the Windows Form Designer.
        InitializeComponent()

        'Add any initialization after the InitializeComponent() call

    End Sub

    'Form overrides dispose to clean up the component list.
    Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
        If disposing Then
            If Not (components Is Nothing) Then
                components.Dispose()
            End If
        End If
        MyBase.Dispose(disposing)
    End Sub

    'Required by the Windows Form Designer
    Private components As System.ComponentModel.IContainer

    'NOTE: The following procedure is required by the Windows Form Designer
    'It can be modified using the Windows Form Designer.  
    'Do not modify it using the code editor.
    Friend WithEvents databaseList As System.Windows.Forms.ComboBox
    Friend WithEvents label5 As System.Windows.Forms.Label
    Friend WithEvents updateBtn As System.Windows.Forms.Button
    Friend WithEvents dataGrid As System.Windows.Forms.DataGrid
    Friend WithEvents tables As System.Windows.Forms.ComboBox
    Friend WithEvents connectBtn As System.Windows.Forms.Button
    Friend WithEvents password As System.Windows.Forms.TextBox
    Friend WithEvents label3 As System.Windows.Forms.Label
    Friend WithEvents userid As System.Windows.Forms.TextBox
    Friend WithEvents label2 As System.Windows.Forms.Label
    Friend WithEvents server As System.Windows.Forms.TextBox
    Friend WithEvents label1 As System.Windows.Forms.Label
    Friend WithEvents label4 As System.Windows.Forms.Label
    <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()
        Me.databaseList = New System.Windows.Forms.ComboBox
        Me.label5 = New System.Windows.Forms.Label
        Me.updateBtn = New System.Windows.Forms.Button
        Me.dataGrid = New System.Windows.Forms.DataGrid
        Me.tables = New System.Windows.Forms.ComboBox
        Me.connectBtn = New System.Windows.Forms.Button
        Me.password = New System.Windows.Forms.TextBox
        Me.label3 = New System.Windows.Forms.Label
        Me.userid = New System.Windows.Forms.TextBox
        Me.label2 = New System.Windows.Forms.Label
        Me.server = New System.Windows.Forms.TextBox
        Me.label1 = New System.Windows.Forms.Label
        Me.label4 = New System.Windows.Forms.Label
        CType(Me.dataGrid, System.ComponentModel.ISupportInitialize).BeginInit()
        Me.SuspendLayout()
        '
        'databaseList
        '
        Me.databaseList.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList
        Me.databaseList.Location = New System.Drawing.Point(80, 80)
        Me.databaseList.Name = "databaseList"
        Me.databaseList.Size = New System.Drawing.Size(296, 21)
        Me.databaseList.TabIndex = 24
        '
        'label5
        '
        Me.label5.Location = New System.Drawing.Point(8, 88)
        Me.label5.Name = "label5"
        Me.label5.Size = New System.Drawing.Size(64, 16)
        Me.label5.TabIndex = 23
        Me.label5.Text = "Databases"
        '
        'updateBtn
        '
        Me.updateBtn.Anchor = CType((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.updateBtn.Location = New System.Drawing.Point(464, 104)
        Me.updateBtn.Name = "updateBtn"
        Me.updateBtn.Size = New System.Drawing.Size(80, 23)
        Me.updateBtn.TabIndex = 22
        Me.updateBtn.Text = "Update"
        '
        'dataGrid
        '
        Me.dataGrid.Anchor = CType((((System.Windows.Forms.AnchorStyles.Top Or System.Windows.Forms.AnchorStyles.Bottom) _
                    Or System.Windows.Forms.AnchorStyles.Left) _
                    Or System.Windows.Forms.AnchorStyles.Right), System.Windows.Forms.AnchorStyles)
        Me.dataGrid.DataMember = ""
        Me.dataGrid.HeaderForeColor = System.Drawing.SystemColors.ControlText
        Me.dataGrid.Location = New System.Drawing.Point(8, 136)
        Me.dataGrid.Name = "dataGrid"
        Me.dataGrid.Size = New System.Drawing.Size(536, 384)
        Me.dataGrid.TabIndex = 21
        '
        'tables
        '
        Me.tables.DropDownStyle = System.Windows.Forms.ComboBoxStyle.DropDownList
        Me.tables.Location = New System.Drawing.Point(80, 104)
        Me.tables.Name = "tables"
        Me.tables.Size = New System.Drawing.Size(296, 21)
        Me.tables.TabIndex = 20
        '
        'connectBtn
        '
        Me.connectBtn.Location = New System.Drawing.Point(400, 8)
        Me.connectBtn.Name = "connectBtn"
        Me.connectBtn.TabIndex = 19
        Me.connectBtn.Text = "Connect"
        '
        'password
        '
        Me.password.Location = New System.Drawing.Point(264, 32)
        Me.password.Name = "password"
        Me.password.PasswordChar = Microsoft.VisualBasic.ChrW(42)
        Me.password.Size = New System.Drawing.Size(116, 20)
        Me.password.TabIndex = 18
        Me.password.Text = ""
        '
        'label3
        '
        Me.label3.Location = New System.Drawing.Point(192, 40)
        Me.label3.Name = "label3"
        Me.label3.Size = New System.Drawing.Size(56, 16)
        Me.label3.TabIndex = 17
        Me.label3.Text = "Password:"
        '
        'userid
        '
        Me.userid.Location = New System.Drawing.Point(56, 32)
        Me.userid.Name = "userid"
        Me.userid.Size = New System.Drawing.Size(120, 20)
        Me.userid.TabIndex = 16
        Me.userid.Text = ""
        '
        'label2
        '
        Me.label2.Location = New System.Drawing.Point(8, 40)
        Me.label2.Name = "label2"
        Me.label2.Size = New System.Drawing.Size(48, 16)
        Me.label2.TabIndex = 15
        Me.label2.Text = "User Id:"
        '
        'server
        '
        Me.server.Location = New System.Drawing.Point(56, 8)
        Me.server.Name = "server"
        Me.server.Size = New System.Drawing.Size(320, 20)
        Me.server.TabIndex = 14
        Me.server.Text = ""
        '
        'label1
        '
        Me.label1.Location = New System.Drawing.Point(8, 16)
        Me.label1.Name = "label1"
        Me.label1.Size = New System.Drawing.Size(48, 16)
        Me.label1.TabIndex = 12
        Me.label1.Text = "Server:"
        '
        'label4
        '
        Me.label4.Location = New System.Drawing.Point(8, 112)
        Me.label4.Name = "label4"
        Me.label4.Size = New System.Drawing.Size(64, 16)
        Me.label4.TabIndex = 13
        Me.label4.Text = "Tables"
        '
        'Form1
        '
        Me.AutoScaleBaseSize = New System.Drawing.Size(5, 13)
        Me.ClientSize = New System.Drawing.Size(552, 525)
        Me.Controls.Add(Me.databaseList)
        Me.Controls.Add(Me.label5)
        Me.Controls.Add(Me.updateBtn)
        Me.Controls.Add(Me.dataGrid)
        Me.Controls.Add(Me.tables)
        Me.Controls.Add(Me.connectBtn)
        Me.Controls.Add(Me.password)
        Me.Controls.Add(Me.label3)
        Me.Controls.Add(Me.userid)
        Me.Controls.Add(Me.label2)
        Me.Controls.Add(Me.server)
        Me.Controls.Add(Me.label1)
        Me.Controls.Add(Me.label4)
        Me.Name = "Form1"
        Me.Text = "Form1"
        CType(Me.dataGrid, System.ComponentModel.ISupportInitialize).EndInit()
        Me.ResumeLayout(False)

    End Sub

#End Region

    Private Sub connectBtn_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles connectBtn.Click
        If Not conn Is Nothing Then conn.Close()

        Dim connStr As String
        connStr = String.Format("server={0};user id={1}; password={2}; database=mysql; pooling=false", _
				server.Text, userid.Text, password.Text )

        Try
            conn = New MySqlConnection(connStr)
            conn.Open()

            GetDatabases()
        Catch ex As MySqlException
            MessageBox.Show("Error connecting to the server: " + ex.Message)
        End Try
    End Sub

    Private Sub GetDatabases()
        Dim reader As MySqlDataReader
        reader = Nothing

        Dim cmd As New MySqlCommand("SHOW DATABASES", conn)
        Try
            reader = cmd.ExecuteReader()
            databaseList.Items.Clear()

            While (reader.Read())
                databaseList.Items.Add(reader.GetString(0))
            End While
        Catch ex As MySqlException
            MessageBox.Show("Failed to populate database list: " + ex.Message)
        Finally
            If Not reader Is Nothing Then reader.Close()
        End Try

    End Sub

    Private Sub databaseList_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles databaseList.SelectedIndexChanged
        Dim reader As MySqlDataReader

        conn.ChangeDatabase(databaseList.SelectedItem.ToString())

        Dim cmd As New MySqlCommand("SHOW TABLES", conn)

        Try
            reader = cmd.ExecuteReader()
            tables.Items.Clear()

            While (reader.Read())
                tables.Items.Add(reader.GetString(0))
            End While

        Catch ex As MySqlException
            MessageBox.Show("Failed to populate table list: " + ex.Message)
        Finally
            If Not reader Is Nothing Then reader.Close()
        End Try
    End Sub

    Private Sub tables_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles tables.SelectedIndexChanged
        data = New DataTable

        da = New MySqlDataAdapter("SELECT * FROM " + tables.SelectedItem.ToString(), conn)
        cb = New MySqlCommandBuilder(da)

        da.Fill(data)

        dataGrid.DataSource = data
    End Sub

    Private Sub updateBtn_Click(ByVal sender As Object, ByVal e As EventArgs) Handles updateBtn.Click
        Dim changes As DataTable = data.GetChanges()
        da.Update(changes)
        data.AcceptChanges()
    End Sub

End Class
